### PR TITLE
feat: add force_use_at_iteration on actions

### LINKED ISSUE #4597: [Multi-Actions] Figure out generic action configurations
Reference: https://www.notion.so/dust-tt/Multi-Actions-Technical-Proposal-662aa8afbc584f4285d89becc0a6417b

This one requires a bit of framing

The shared action configurations fields are:
- `description`
- `force_use_at_iteration`

Maybe `max_uses_per_run`, `min_uses_per_run`

We need to figure out where we will store that. I presume the simplest is that it will exist on all action configuration models.

As part of this, add these new fields to all existing action configuration (including generation)

Figure out what are the invariants in naming that cover existing actions, generation and future one. What are the base classes in code? BaseActionResource likely?

### PR DESCRIPTION
**Pull request by fontanierh**
## Description

fixes https://github.com/dust-tt/dust/issues/4597

We're adding `forceUseAtIteration` on all our action configurations. This will allow to support "legacy" single action assistants in the new multi-actions model.

this PR adds the field (both in DB and in the types / input schemas) and also adds a new "legacyMode" flag when creating/updating assistants that automatically populates the field with the right values.

There is also a migration script so every actions and generation configs gets its backfilled `forceUseAtIteration`

## Deploy Plan

First, initdb. Then deploy code. Then, run the migration script.

### CODE REVIEW

File: front/migrations/20240412_force_use_at_iteration.ts
Comment (philipperolet) [inline front/migrations/20240412_force_use_at_iteration.ts:66]: did you test that? on my localhost skips all agents

**Review by philipperolet — COMMENTED** [review]
Looks good! Left a few coms, and making last local tests before final approval

File: front/migrations/20240412_force_use_at_iteration.ts
Comment (philipperolet) [reply front/migrations/20240412_force_use_at_iteration.ts:66]: could be because of a weird local state because of all ongoing migrations but worth checking

File: types/src/front/api_handlers/internal/agent_configuration.ts
Comment (fontanierh) [reply types/src/front/api_handlers/internal/agent_configuration.ts:62]: Yes it's the only way to allow omitting the key... Really sucks tbh

File: types/src/front/assistant/agent.ts
Comment (philipperolet) [reply types/src/front/assistant/agent.ts:59]: 3 fields if we count id & sId. 
Since we're going to add many actions soon, good to have the type checker remind people to fill those 3 fields (e.g. I think sId is missing on GenerationConfiguration)

maybe outside scope of the PR but would you think it legitimate to have a separate PR (by me, you or someone else) for that?

File: types/src/front/api_handlers/internal/agent_configuration.ts
Comment (philipperolet) [reply types/src/front/api_handlers/internal/agent_configuration.ts:62]: :nod:

File: front/migrations/20240412_force_use_at_iteration.ts
Comment (philipperolet) [reply front/migrations/20240412_force_use_at_iteration.ts:66]: the thing is it called "skipped because no generationConfig" while they do have one (not related to execute, right?)

File: types/src/front/assistant/agent.ts
Comment (fontanierh) [reply types/src/front/assistant/agent.ts:59]: fair fair, happy to do that

File: front/migrations/20240412_force_use_at_iteration.ts
Comment (fontanierh) [reply front/migrations/20240412_force_use_at_iteration.ts:66]: That is definitely weird

Comment (philipperolet) [reply front/migrations/20240412_force_use_at_iteration.ts:66]: as per irl, should not select only agentconfig with instructions: null :+1:

File: front/migrations/20240412_force_use_at_iteration.ts
Comment (PopDaph) [inline front/migrations/20240412_force_use_at_iteration.ts:64]: Since we want to merge my migration and yours, here we need to add something like: 

```
if (execute) {
    const nbFixedIterations = (generation ? 1 : 0) + actions.length
    await AgentConfiguration.update({ maxToolsUseForRun: 1 }, {
    where: {
      id: aId,
    },
  });
}
````

Comment (PopDaph) [reply front/migrations/20240412_force_use_at_iteration.ts:64]: (wrote from Github so definitely not reliable)

File: front/migrations/20240412_force_use_at_iteration.ts
Comment (fontanierh) [reply front/migrations/20240412_force_use_at_iteration.ts:63]: Thank you for catching 🙏

**Review by philipperolet — APPROVED** [review]
LGTM modulo approval of Daph wrt the consanguine migrations :wink:

File: types/src/front/assistant/agent.ts
Comment (fontanierh) [reply types/src/front/assistant/agent.ts:59]: Let's do it as soon as we move generation to actions so we don't have to do weird imports

File: front/migrations/20240412_force_use_at_iteration.ts
Comment (fontanierh) [reply front/migrations/20240412_force_use_at_iteration.ts:64]: added

**Review by PopDaph — APPROVED** [review]
LGTM

File: front/lib/api/assistant/configuration.ts
Comment (spolu) [inline front/lib/api/assistant/configuration.ts:1044]: Why are objects doubled above?

**Review by spolu — APPROVED** [review]
LGTM

Left a few aesthetics comments

File: types/src/front/assistant/agent.ts
Comment (philipperolet) [reply types/src/front/assistant/agent.ts:59]: seeing as we have 4 actions sharing 3 fields, and will soon have 3 more actions, isn't that making it easier for us to automate these 3 fields? VS mental load for people introducing new actions to add the fields with the right consistency everywhere needed? 

If we're sure we're going to add many actions, and will want to treat them partly similarly, a common ground seems to makes sense?

(just pulling the argument a bit, won't fight harder than that for this)

Comment (spolu) [reply types/src/front/assistant/agent.ts:59]: Factorising actions is a much broader project and we’ll need to think hard about how we want to do it (likely using TS classes and inheritance to have all implementation of an action in the same place + generic code for interaction with it in our conversations logic)

File: front/lib/api/assistant/configuration.ts
Comment (fontanierh) [reply front/lib/api/assistant/configuration.ts:1044]: I'm not quite sure what you mean here ?

**Pull request merged**
